pkg/apis/jaegertracing/v1/deployment_strategy.go

```diff
@@ -0,0 +1,49 @@
+package v1
+
+import (
+	"errors"
+	"strings"
+)
+
+// DeploymentStrategy represents the possible values for deployment strategies
+// +k8s:openapi-gen=true
+type DeploymentStrategy string
+
+const (
+	// DeploymentStrategyDeprecatedAllInOne represents the (deprecated) 'all-in-one' deployment strategy
+	// +k8s:openapi-gen=true
+	DeploymentStrategyDeprecatedAllInOne DeploymentStrategy = "all-in-one"
+
+	// DeploymentStrategyAllInOne represents the 'allInOne' deployment strategy (default)
+	// +k8s:openapi-gen=true
+	DeploymentStrategyAllInOne DeploymentStrategy = "allinone"
+
+	// DeploymentStrategyStreaming represents the 'streaming' deployment strategy
+	// +k8s:openapi-gen=true
+	DeploymentStrategyStreaming DeploymentStrategy = "streaming"
+
+	// DeploymentStrategyProduction represents the 'production' deployment strategy
+	// +k8s:openapi-gen=true
+	DeploymentStrategyProduction DeploymentStrategy = "production"
+)
+
+// UnmarshalText implements encoding.TextUnmarshaler to ensure that JSON values in the
+// strategy field of JSON jaeger specs are interpreted in a case-insensitive manner
+func (ds *DeploymentStrategy) UnmarshalText(text []byte) error {
+	if ds == nil {
```

Comment:
This is interesting. I'm probably missing something obvious, but how would this be triggered? Wouldn't a nil object fail with a nil reference before reaching this code? Can you come up with a test case for this?

Reply:
It's not obvious. It is possible for the function to be called on a nil receiver when interfaces are involved (as is the case here).

I was digging into the docs and I had a look at how [json.RawMessage.UnmarshalJSON](https://golang.org/src/encoding/json/stream.go?s=6952:7005#L257) is implemented. The check for nil felt strange, so I did a bit of research and came up with this https://tour.golang.org/methods/12

Wonderful Go! :)

Reply:
BTW, I guess this kind of check should be implemented in other types implementing interfaces for additional safety. `Options` come to my mind, implementing both `json.Marshaler` and `json.Unmarshaler` interfaces, but I'm sure there are several others.

Reply:
TIL! I'm merging this one. Would you mind creating an issue to do the same nil check in the other relevant types?

Reply:
Will do!